runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -0,0 +1,367 @@
+mod bpf_upgradeable;
+mod builtin;
+pub(crate) mod error;
+
+use {
+    crate::bank::Bank,
+    bpf_upgradeable::SourceProgramBpfUpgradeable,
+    builtin::TargetProgramBuiltin,
+    error::CoreBpfMigrationError,
+    solana_sdk::{
+        account::{AccountSharedData, ReadableAccount},
+        bpf_loader_upgradeable::UpgradeableLoaderState,
+        pubkey::Pubkey,
+    },
+    std::sync::atomic::Ordering::Relaxed,
+};
+
+/// Sets up a Core BPF migration for a built-in program.
+#[derive(Debug)]
+pub enum CoreBpfMigration {
```

Comment:
nit: in the end, this defines the target program, so could we rename this to `CoreBpfTargetType` or something?